File: ruby/bob/bob.rb
require 'pry'

class Bob
  attr_reader :remark

  def hey(remark)
    @remark = remark

    case
    when silence? then "Fine. Be that way!"
    when shouting? then "Whoa, chill out!"
    when question? then "Sure."
    else "Whatever."
    end
  end

  def silence?
    remark.strip.empty?
  end

  def question?
    remark.end_with?("?")
  end

  def shouting?
    non_digits = remark.match(/[A-Za-z]+/).to_s
    binding.pry
    non_digits == non_digits.upcase unless non_digits.empty?
  end
end

File: csharp/bob/Bob.cs
using System;
using System.Linq;

public static class Bob
{
    private static class Responses {
        public static string Question = "Sure.";
        public static string Shouting = "Whoa, chill out!";
        public static string Indifferent = "Whatever.";
        public static string Nonesense = "Fine. Be that way!";
        public static string ShoutingQuestion = "Calm down, I know what I'm doing!";
    }

    public static string Response(string statement)
    {
        statement = statement.Trim();

        if(statement == "") { return Responses.Nonesense; }        

        if(AllCaps(statement) && EndsWithQuestion(statement)) { return Responses.ShoutingQuestion; }
        if(AllCaps(statement)) { return Responses.Shouting; }
        if(EndsWithQuestion(statement)) { return Responses.Question; }
        if(Empty(statement)) { return Responses.Nonesense; }

        return Responses.Indifferent;
    }

    private static bool EndsWithQuestion(string statement)
    {
        return statement[statement.Length - 1] == '?';
    }

    private static bool Empty(string statement)
    {
        foreach(char i in statement) {
            if(char.IsLetterOrDigit(i)) {
                return false;
            }
        }

        return true;
    }

    private static bool AllCaps(string statement)
    {
        string testStatement = "";
        foreach(char i in statement) {
            if(char.IsLetter(i)) {
                testStatement += i;
            }
        }

        if(testStatement == ""){
            return false;
        }

        return testStatement.All(char.IsUpper);
    }
}